### PR TITLE
Add helpers for checking whether a transaction must be added to the limbo and whether a transaction needs to be acked

### PR DESCRIPTION
**Pull request by CuriousGeorgiy**
This patch series adds helpers for checking whether a transaction must be added to the limbo and whether a transaction needs to be acked.

Needed for tarantool/tarantool-ee#752
Needed for tarantool/tarantool-ee#755

### CODE REVIEW

**Review by Serpentian — CHANGES_REQUESTED** [review]
Thank you for the patch, but I'd rather merge it alongside with the main one. Now it's easy to miss some places and IMHO, it'll look confusing in master and will complicate the existing code.

File: src/box/txn.h [hunk 1]
```diff
@@ -534,6 +534,12 @@ txn_has_flag(const struct txn *txn, enum txn_flag flag)
 	return (txn->flags & flag) != 0;
 }
 
+static inline bool
+txn_needs_ack(const struct txn *txn)
+{
+	return (txn->flags & TXN_WAIT_ACK) != 0;
```
[the remark this answers is not in the document]
Comment (CuriousGeorgiy) [reply src/box/txn.h:546]: I wrote it this way to extend it for `TXN_EARLY_ACK` like this: `return (txn->flags & (TXN_WAIT_ACK | TXN_EARLY_ACK)) != 0;`.

File: src/box/txn.c [hunk 2]
```diff
@@ -1090,7 +1090,7 @@ txn_commit_try_async(struct txn *txn)
 	if (req == NULL)
```
[the remark this answers is not in the document]
Comment (CuriousGeorgiy) [reply src/box/txn.c:1090]: Yeah, sorry, my bad, replaced it everywhere.

File: src/box/txn.c [hunk 3]
```diff
@@ -1155,7 +1155,7 @@ txn_commit(struct txn *txn)
 		 * another WAL write. Can't be done until it works
```
[the remark this answers is not in the document]
Comment (CuriousGeorgiy) [reply src/box/txn.c:1155]: This place below boggled me:
https://github.com/tarantool/tarantool/blob/4cac1677c88a0ee13944d8dc44b10813738bb445/src/box/txn.c#L940-L950

I think we can simply add a check like this below:
```c++
if (!txn_needs_ack(txn))
	txn_set_flags(txn, TXN_WAIT_ACK);
txn_set_flags(txn, TXN_WAIT_SYNC);
```

File: src/box/txn.c [hunk 4]
```diff
@@ -1090,7 +1090,7 @@ txn_commit_try_async(struct txn *txn)
 	if (req == NULL)
 		goto rollback;
```
[the remark this answers is not in the document]
Comment (CuriousGeorgiy) [reply src/box/txn.c:1093]: Fixed.

File: src/box/txn.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (Serpentian) [reply src/box/txn.c:1157]: I'd rather left it as it is. Why do we need this additional condition here?

File: src/box/txn.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (Serpentian) [reply src/box/txn.c:1093]: It's still not there, AFAICS)

File: src/box/txn.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (CuriousGeorgiy) [reply src/box/txn.c:1157]: According to the RFC [^1], we should not always add the `TXN_WAIT_ACK` flag in this place:

> Given the new flag the possible flag combinations for limbo entries are are:
> 
> - `TXN_EARLY_ACK` only: for async transactions in the limbo which don’t have any sync > transactions before them (thus these async transactions are already ACK-ed to the user)
> - `TXN_WAIT_ACK + TXN_WAIT_SYNC` : for synchronous transactions
> - `TXN_WAIT_SYNC + TXN_EARLY_ACK`: for async transactions in the limbo after some sync transaction (thus these transactions were not yet ACKed to the user). As soon as the preceding sync transaction gets confirmed, these transactions are ACKed to the user, and their flag is changed to `TXN_EARLY_ACK` (`TXN_WAIT_SYNC` is removed). At the same time they become the first ones in the queue.

If an asynchronous transaction came, we should only add the `TXN_WAIT_SYNC` flag to it. Not sure about the best way to handle this logic here.

[^1]: https://www.notion.so/tarantool/Async-in-the-presence-of-sync-9456c50afefc4b2dbe5514d8e1582803?pvs=4#f3adc2554f414e599186a268e0f9e429

File: src/box/txn.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (CuriousGeorgiy) [reply src/box/txn.c:1093]: Sorry, the rewording got dropped for some reason, should be fine now.

File: src/box/txn.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (Serpentian) [reply src/box/txn.c:1157]: Can we leave this place untouched for now? It looks strange without the main patch and interferes with code readability:

```c
if (!txn_has_flag(txn,  TXN_WAIT_ACK)
    txn_set_flags(txn, TXN_WAIT_ACK)
txn_set_flags(txn, TXN_WAIT_SYNC)
```

**Review by sergepetrenko — COMMENTED** [review]
I'd rather not move on with this refactoring until we see the full picture with the main patch.

For example, instead of introducing helpers like `txn_must_be_in_limbo`, which replaces `TXN_EARLY_ACK | TXN_WAIT_SYNC`, you could rearrange the txn flags and introduce a separate `TXN_MUST_BE_IN_LIMBO` flag.

Same for `txn_needs_ack`.

So the flags would be:
`TXN_MUST_BE_IN_LIMO` (feel free to find a better name).
`TXN_NEEDS_ACK`
`TXN_EARLY_ACK`.

Then `TXN_MUST_BE_IN_LIMBO` is set for all 3 types of transactions: async-before-sync, sync, and async-after-sync transactions.
`TXN_NEEDS_ACK`  is set for async-before-sync and sync ones.
`TXN_EARLY_ACK` is set for async-before-sync ones.

Alternatively you could remove `TXN_EARLY_ACK` altogether and instead introduce `TXN_IS_SYNC` which is set only for the sync transactions.

Anyway, if predicates `txn_must_be_in_limbo`/`txn_needs_ack` become identical to checking a single flag like `txn_has_flag(txn, TXN_MUST_BE_IN_LIMBO)`, I think checking the flag directly would be more straightforward.

File: src/box/txn.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (CuriousGeorgiy) [reply src/box/txn.c:1157]: Thanks for the feedback! Closing this PR in the light of #10461.

**Comment by CuriousGeorgiy** [comment]
Thanks for the feedback! Closing this PR in the light of #10461.

**Pull request closed without merging**